File: xiaqianZhang/assignments/heaps/lc295/lc295.py
# Problem Statement #

# Design a class to calculate the median of a number stream. The class should have the following two methods:

#   1. insertNum(int num): stores the number in the class
#   2. findMedian(): returns the median of all numbers inserted in the class

# If the count of numbers inserted in the class is even, the median will be the average of the middle two numbers.


'''
class MedianFinder:
    def __init__(self):
        self.small = []  # the smaller half of the list, max heap (invert min-heap)
        self.large = []  # the larger half of the list, min heap

    def addNum(self, num):
        if len(self.small) == len(self.large):
            heappush(self.large, -heappushpop(self.small, -num))
        else:
            heappush(self.small, -heappushpop(self.large, num))

    def findMedian(self):
        if len(self.small) == len(self.large):
            return float(self.large[0] - self.small[0]) / 2.0
        else:
            return float(self.large[0])

large = [1]
num = 2
large = [1, 2] root = 1 since it is min heap
pop out the -1
num = -1
push into the small array

small = [-1]
large = [2]

num = 3
small [-1, -3] -3 pop out since it is min heap
after the pop out, we reversed the sign and push to the large 
small [-1]
large [2, 3]

num = 4
large [2, 3, 4]
pop out the min heap, pop out 2
push -2 into the small
small [-2, -1]
large [3, 4]

num = 5
push the -5 into small array
small [-5, -2, -1]
pop out the -5
push -(-5) into large subarray

large [3, 4, 5]

'''

'''
Attempt 2

Idea: create two heap, 
one is min heap that contains the larger numbers
one is max heap that contains the smaller numbers
Then the root of the min heap is the largest value within that heap
And the root of the max heap is the largest value within this heap
Then check if these two heap have the same length size
if so, take two heap's root value and divide by 2
else, take the root value of the min heap side
'''
from heapq import *

class MedianOfAStream:
  def __init__(self):
    self.smaller = []
    self.larger = []

  def insert_num(self, num):
    if len(self.smaller) == len(self.larger):
      heappush(self.larger, -heappushpop(self.smaller, -num))
    else:
      heappush(self.smaller, -heappushpop(self.larger, num))

  def find_median(self):
    if len(self.smaller) == len(self.larger):
      return (self.larger[0] - self.smaller[0]) / 2
    else:
      return self.larger[0]

# from heapq import *

# class MedianOfAStream:
#   def __init__(self):
#     self.small = []
#     self.large = []

#   def insert_num(self, num):
#     if len(self.small) == len(self.large):
#       heappush(self.large, -heappushpop(self.small, -num))
#     else:
#       heappush(self.small, -heappushpop(self.large, num))


#   def find_median(self):
#     if len(self.small) == len(self.large):
#       #the root of each side and divide by 2
#       return (-self.small[0] + self.large[0]) / 2
#     else:
#       return self.large[0]


def main():
  medianOfAStream = MedianOfAStream()
  # []
  medianOfAStream.insert_num(3)
  # [3]
  medianOfAStream.insert_num(1)
  # [1, 3]
  print("The median is: " + str(medianOfAStream.find_median()))
  # 2
  medianOfAStream.insert_num(5)
  print("The median is: " + str(medianOfAStream.find_median()))
  medianOfAStream.insert_num(4)
  # index 4 -1 // 2, so the parent is 5, swap becomes [1, 3, 4, 5]
  print("The median is: " + str(medianOfAStream.find_median()))


main()

File: janMarkusMilan/assignments/heaps/lc295/lc295.js
// Problem Statement #

// Design a class to calculate the median of a number stream. The class should have the following two methods:

//   1. insertNum(int num): stores the number in the class
//   2. findMedian(): returns the median of all numbers inserted in the class

// If the count of numbers inserted in the class is even, the median will be the average of the middle two numbers.

/*
Understanding the problem:
  1. Given some array, we need to find the median
  2. To get the median, the values in the array must first be sorted in ascending order
  3. We need to implement the actual methods for the class

Constraints:
  1. The values in the array can be positive or negative
  2. The length of the array will, at minimum, be one before calling find_median()

Edge Cases:
  - Little to none. We will always be returning some value, 
    as there will be at least one value in the array when finding the median.

Understanding heaps:
  - Let's say we were given an array: [1, 2, 3, 4, 5, 6]
    We would split the array in half: [1, 2, 3][4, 5, 6]
  - Essentially, we can put the smaller half into a max heap, the the bigger half into a min heap.
    That way, we can get the middle values in constant time. 

    (3) <---        (4) <---
    / \             / \
  (1) (2)         (5) (6)

Break the problem into small sub-problems:
  1. Create a class Heap with the methods:
      - constructor
      - add method
      - peek method
      - poll method
      - bubbleUp method
      - bubbleDown method
  2. After creating the Heap class, implement the insert_num method
  3. Finally, implement the find_median method  

-------------------------------------------------------

Median's addNum method (pseudo-code)
-----
If no values in maxHeap OR num passed into addNum is less than topValue of MaxHeap:
    Add this value to the maxHeap
Else
    Add it to the minHeap
    
If maxHeap is growing faster than minHeap
    Pull top values with poll and add to opposing heap
If OPPOSITE
    OPPOSITE
    
-------------------------------------------------------
    
Median's findMedian method (pseudo-code)
-----
If maxHeap has more values than minHeap:
    median is here in maxHeap
Else if OPPOSITE:
    median is here in minHeap
Else:
    even number so we need to do math in order to get median  
    
-------------------------------------------------------
*/

class MedianOfAStream {
  constructor() {
    this.maxHeap = new Heap(Heap.maxComparator);
    this.minHeap = new Heap(Heap.minComparator);
  }

  insert_num(num) {
    // TODO: Write your code here
    if (this.maxHeap.peek() === null || num < this.maxHeap.peek()) {
      this.maxHeap.add(num);
    } else {
      this.minHeap.add(num);
    }

    if (this.maxHeap.size - this.minHeap.size > 1) {
      this.minHeap.add(this.maxHeap.poll());
    } else if (this.minHeap.size - this.maxHeap.size > 1) {
      this.maxHeap.add(this.minHeap.poll());
    }
  }

  find_median(self) {
    // TODO: Write your code here
    if (this.maxHeap.size > this.minHeap.size) {
      return this.maxHeap.peek();
    } else if (this.minHeap.size > this.maxHeap.size) {
      return this.minHeap.peek();
    } else {
      return (this.minHeap.peek() + this.maxHeap.peek()) / 2;
    }
  }
}

class Heap {
  constructor(comparator) {
    this.size = 0;
    this.values = [];
    this.comparator = comparator || Heap.minComparator;
  }

  add(val) {
    this.values.push(val);
    this.size++;
    this.bubbleUp();
  }

  peek() {
    return this.values[0] || null;
  }

  poll() {
    const max = this.values[0];
    const end = this.values.pop();
    this.size--;

    if (this.values.length) {
      this.values[0] = end;
      this.bubbleDown();
    }

    return max;
  }

  bubbleUp() {
    let index = this.values.length - 1;
    let parent = Math.floor((index - 1) / 2);

    while (this.comparator(this.values[index], this.values[parent]) < 0) {
      [this.values[parent], this.values[index]] = [
        this.values[index],
        this.values[parent],
      ];

      index = parent;
      parent = Math.floor((index - 1) / 2);
    }
  }

  bubbleDown() {
    let index = 0;
    let length = this.values.length;

    while (true) {
      let left = null;
      let right = null;
      let swap = null;
      let leftIndex = index * 2 + 1;
      let rightIndex = index * 2 + 2;

      if (leftIndex < length) {
        left = this.values[leftIndex];
        if (this.comparator(left, this.values[index]) < 0) {
          swap = leftIndex;
        }
      }

      if (rightIndex < length) {
        right = this.values[rightIndex];
        if (
          (swap !== null && this.comparator(right, left) < 0) ||
          (swap === null && this.comparator(right, this.values[index]))
        ) {
          swap = rightIndex;
        }
      }

      if (swap === null) {
        break;
      }

      [this.values[index], this.values[swap]] = [
        this.values[swap],
        this.values[index],
      ];
      index = swap;
    }
  }
}

Heap.minComparator = (a, b) => {
  return a - b;
};
Heap.maxComparator = (a, b) => {
  return b - a;
};

var medianOfAStream = new MedianOfAStream();
medianOfAStream.insert_num(3);
medianOfAStream.insert_num(1);
console.log(`The median is: ${medianOfAStream.find_median()}`);
medianOfAStream.insert_num(5);
console.log(`The median is: ${medianOfAStream.find_median()}`);
medianOfAStream.insert_num(4);
console.log(`The median is: ${medianOfAStream.find_median()}`);
